File: zeroMQ/python/receiver_async_test.py
#!/usr/bin/env python3
"""ZeroMQ Python Receiver - Async"""
import sys
import signal
import asyncio
import zmq
import zmq.asyncio
from pathlib import Path

# Add utils to path
repo_root = Path(__file__).parent.parent.parent
sys.path.insert(0, str(repo_root / 'utils' / 'python'))

from message_helpers import *

running = True

def signal_handler(sig, frame):
    global running
    running = False

async def run(receiver_id):
    context = zmq.asyncio.Context()
    socket = context.socket(zmq.REP)
    port = 5556 + receiver_id
    socket.bind(f"tcp://*:{port}")
    
    print(f" [*] [ASYNC] Receiver {receiver_id} listening on port {port}")
    
    while running:
        if await socket.poll(100):
            message = await socket.recv()
            
            request_envelope = parse_envelope(message)
            message_id = request_envelope.message_id
            print(f" [x] [ASYNC] Received message {message_id}")
            
            # Create ACK
            response = create_ack_from_envelope(request_envelope, str(receiver_id))
            setattr(response, 'async', True)
            resp_str = serialize_envelope(response)
            
            await socket.send(resp_str)
        else:
            # Yield control to allow loop to process signals/other tasks
            await asyncio.sleep(0.01)
            
    print(f" [x] [ASYNC] Receiver {receiver_id} shutting down")
    socket.close()
    context.term()


def main():
    import argparse
    parser = argparse.ArgumentParser()
    parser.add_argument('--id', type=int, default=0)
    args = parser.parse_args()
    
    signal.signal(signal.SIGINT, signal_handler)
    signal.signal(signal.SIGTERM, signal_handler)
    
    asyncio.run(run(args.id))


if __name__ == "__main__":
    main()

File: zeroMQ/cpp/receiver_async_test.cpp
#include <zmq.hpp>
#include <iostream>
#include <string>
#include <signal.h>
#include <atomic>
#include <thread>
#include "../../utils/cpp/message_helpers.hpp"

using messaging::MessageEnvelope;
using message_helpers::get_current_time_ms;

std::atomic<bool> running(true);

void signal_handler(int sig) {
    running = false;
}

int main(int argc, char* argv[]) {
    int receiver_id = 0;
    
    for (int i = 1; i < argc; i++) {
        if (std::string(argv[i]) == "--id" && i + 1 < argc) {
            receiver_id = std::stoi(argv[i + 1]);
            break;
        }
    }
    
    signal(SIGINT, signal_handler);
    signal(SIGTERM, signal_handler);

    zmq::context_t context(1);
    zmq::socket_t socket(context, ZMQ_REP);
    
    int port = 5556 + receiver_id;
    socket.bind("tcp://*:" + std::to_string(port));
    socket.setsockopt(ZMQ_RCVTIMEO, 1000);  // 1s timeout for graceful shutdown

    std::cout << " [*] [ASYNC] Receiver " << receiver_id << " listening on port " << port << std::endl;

    while (running) {
        zmq::message_t request;
        auto recv_res = socket.recv(request);
        
        if (recv_res.has_value()) {
            std::string request_str(static_cast<char*>(request.data()), request.size());
            
            // Parse message
            MessageEnvelope msg_envelope;
            if (message_helpers::parse_envelope(request_str, msg_envelope)) {
                std::string message_id = msg_envelope.message_id();
                std::cout << " [x] [ASYNC] Received message " << message_id << std::endl;
                
                // Create ACK
                MessageEnvelope response = message_helpers::create_ack_from_envelope(
                    msg_envelope,
                    std::to_string(receiver_id)
                );
                response.set_async(true);
                std::string response_str = message_helpers::serialize_envelope(response);
                
                // Send ACK
                zmq::message_t reply(response_str.size());
                memcpy(reply.data(), response_str.c_str(), response_str.size());
                socket.send(reply, zmq::send_flags::none);
            }
        }
    }

    std::cout << " [x] [ASYNC] Receiver " << receiver_id << " shutting down" << std::endl;
    socket.close();
    return 0;
}
